### ISSUE TITLE
how much memory are needed at least for training a 14B-parameter model?

### ISSUE THREAD

**Issue by hsjkdjj** [NVlabs/rcm#12]
hello,  What is the minimum memory requirement and how many machines are needed at least for training a 14B-parameter model?

**Comment by zhengkw18** [comment]
The current training infrastructure is not good enough, so the CP size must be divisible by the number of latent frames (20). If you have GPUs whose number is a multiple of 10, you can use CP size 10 for now. We are working on a better training infra to support more flexible CP size.

**Comment by zhengkw18** [comment]
If you want to train with 8 GPUs for now, you can decrease the `model.config.state_t` to 8 or 16 to train on shorter videos, and set CP size to 8. The number of 8-GPU machines is the batch size, better to use a larger batch size.

**Comment by renrenzsbbb** [comment]
> If you want to train with 8 GPUs for now, you can decrease the `model.config.state_t` to 8 or 16 to train on shorter videos, and set CP size to 8. The number of 8-GPU machines is the batch size, better to use a larger batch size.

soory, even when I decrease state_t to 16 and set CP size to 8, It still meeting with CUDA OOM on   8-GPU A100 machines. Can you introduce me other way to fix it? can I remove the encdoer and decoder part in below code
`            if self.input_latent_key in data_batch:
                self._normalize_latent_inplace(data_batch)
                data_batch[self.input_data_key] = self.decode(data_batch[self.input_latent_key]).contiguous().float().clamp(-1, 1)
                data_batch[IS_PREPROCESSED_KEY] = True

            
            self._normalize_video_inplace(data_batch) # sample in temporal dimension
            data_batch[self.input_latent_key] = self.encode(data_batch[self.input_data_key]).contiguous().float()
            data_batch[IS_PROCESSED_KEY] = True
`

**Comment by renrenzsbbb** [comment]
1. I only can use 8*A100(80G), which have same gpu memory as H100;
2. I set state_t to 8 which still cause cuda OOM;
3. I try it both "mm_only"  and "block_wise". which all cause cuda OOM. by the way, I think mm_only will save memory rather than block_wise, can you give me more suggestion.

**Comment by renrenzsbbb** [comment]
> > 1. I only can use 8*A100(80G), which have same gpu memory as H100;
> > 2. I set state_t to 8 which still cause cuda OOM;
> > 3. I try it both "mm_only"  and "block_wise". which all cause cuda OOM. by the way, I think mm_only will save memory rather than block_wise, can you give me more suggestion.
> 
> After investigation, I find that 8 GPUs with 80G memory are not enough for 14B training. This is because DMD requires 2 optimizers; without large FSDP sharding, the optimizer states themselves almost occupy all the memory, so even with very small `state_t` it will still OOM.
> 
> Therefore, for 14B training, we recommend using more nodes to allow a larger FSDP sharding size.

Thanks for your reply. it can train with 8*80G by setting
model_parallel.context_parallel_size=8 \
model.config.state_t=8 \
by the way, I find that you update code about CP in turbodiffusion repo, can it reduce gpu memory? if yes, can I train it with more context_parallel_size in 8*80G A100 gpu

**Comment by zhengkw18** [comment]
> > > 1. I only can use 8*A100(80G), which have same gpu memory as H100;
> > > 2. I set state_t to 8 which still cause cuda OOM;
> > > 3. I try it both "mm_only"  and "block_wise". which all cause cuda OOM. by the way, I think mm_only will save memory rather than block_wise, can you give me more suggestion.
> > 
> > 
> > After investigation, I find that 8 GPUs with 80G memory are not enough for 14B training. This is because DMD requires 2 optimizers; without large FSDP sharding, the optimizer states themselves almost occupy all the memory, so even with very small `state_t` it will still OOM.
> > Therefore, for 14B training, we recommend using more nodes to allow a larger FSDP sharding size.
> 
> Thanks for your reply. it can train with 8 * 80G by setting model_parallel.context_parallel_size=8 model.config.state_t=8 by the way, I find that you update code about CP in [Support flattened CP](https://github.com/NVlabs/rcm/commit/8f78998892251e507acbc5454de64595ea05f4d3), can it reduce gpu memory? if yes, can I train it with more context_parallel_size in 8 * 80G A100 gpu

It cannot reduce GPU memory, but aims to support more flexible CP size. For example, you can currently use CP size=8 for state_t=21. Moreover, we recommend using the latest training code because it fixed a previous bug in computing the normalizing factor under CP.

**Comment by renrenzsbbb** [comment]
> > > > 1. I only can use 8*A100(80G), which have same gpu memory as H100;
> > > > 2. I set state_t to 8 which still cause cuda OOM;
> > > > 3. I try it both "mm_only"  and "block_wise". which all cause cuda OOM. by the way, I think mm_only will save memory rather than block_wise, can you give me more suggestion.
> > > 
> > > 
> > > After investigation, I find that 8 GPUs with 80G memory are not enough for 14B training. This is because DMD requires 2 optimizers; without large FSDP sharding, the optimizer states themselves almost occupy all the memory, so even with very small `state_t` it will still OOM.
> > > Therefore, for 14B training, we recommend using more nodes to allow a larger FSDP sharding size.
> > 
> > 
> > Thanks for your reply. it can train with 8 * 80G by setting model_parallel.context_parallel_size=8 model.config.state_t=8 by the way, I find that you update code about CP in [Support flattened CP](https://github.com/NVlabs/rcm/commit/8f78998892251e507acbc5454de64595ea05f4d3), can it reduce gpu memory? if yes, can I train it with more context_parallel_size in 8 * 80G A100 gpu
> 
> It cannot reduce GPU memory, but aims to support more flexible CP size. For example, you can currently use CP size=8 for state_t=21. Moreover, we recommend using the latest training code because it fixed a previous bug in computing the normalizing factor under CP.

Thank for your reply. Are you referring to the bug in the commit "update warmup"? I have adjust your code to my project, it take more time for me to merge your code.

**Comment by zhengkw18** [comment]
> > > > > 1. I only can use 8*A100(80G), which have same gpu memory as H100;
> > > > > 2. I set state_t to 8 which still cause cuda OOM;
> > > > > 3. I try it both "mm_only"  and "block_wise". which all cause cuda OOM. by the way, I think mm_only will save memory rather than block_wise, can you give me more suggestion.
> > > > 
> > > > 
> > > > After investigation, I find that 8 GPUs with 80G memory are not enough for 14B training. This is because DMD requires 2 optimizers; without large FSDP sharding, the optimizer states themselves almost occupy all the memory, so even with very small `state_t` it will still OOM.
> > > > Therefore, for 14B training, we recommend using more nodes to allow a larger FSDP sharding size.
> > > 
> > > 
> > > Thanks for your reply. it can train with 8 * 80G by setting model_parallel.context_parallel_size=8 model.config.state_t=8 by the way, I find that you update code about CP in [Support flattened CP](https://github.com/NVlabs/rcm/commit/8f78998892251e507acbc5454de64595ea05f4d3), can it reduce gpu memory? if yes, can I train it with more context_parallel_size in 8 * 80G A100 gpu
> > 
> > 
> > It cannot reduce GPU memory, but aims to support more flexible CP size. For example, you can currently use CP size=8 for state_t=21. Moreover, we recommend using the latest training code because it fixed a previous bug in computing the normalizing factor under CP.
> 
> Thank for your reply. Are you referring to the bug in the commit "update warmup"? I have adjust your code to my project, it take more time for me to merge your code.

Warm-up is not very important. The most critical bug happens when normalizing (e.g., https://github.com/NVlabs/rcm/blob/main/rcm/models/t2v_model_distill_rcm.py#L580). In the previous code, all B_C_T_H_W shape tensors in the training step after network output are actually CP-splitted, so the normalizing operation happens separately on each splitted chunks, which is not expected. In the updated code, we let the network return full-shape tensors by concatenating CP-splitted tensors (https://github.com/NVlabs/rcm/blob/main/rcm/networks/wan2pt1.py#L705).

**Comment by wzr0108** [comment]
Did you eventually succeed in training the 14B model? I ran into an OOM (out-of-memory) issue as well.
